Update contributing guidelines

leotrs opened the pull request:
Updated contributing guidelines. Note that this affects the `CONTRIBUTING.md` file as well as the `docs/source/contributing.md` file, since the latter is just a symlink of the former.

Aathish04 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
-   git remote add fork https://github.com/ManimCommunity/manim.git
-   git fetch fork
-   git checkout -b <your-branch-name> fork/master
+   git remote add upstream https://github.com/ManimCommunity/manim.git
+   git pull upstream master
```
```suggestion
   git remote add community https://github.com/ManimCommunity/manim.git
   git pull community master
```

Nitpick: I think naming the remote `community` over `upstream` will provide more clarity as to what exactly the remote is.

leotrs replied:
I disagree here. I think upstream is more standard (and it's also shorter).

Aathish04 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
-    Now there are three repositories that git is keeping track of: the manim community repo (referred to as "fork"), your own fork of it (referred to by git as "origin"), and your local repository.
+	After these commands, your local repository can keep track of your fork
+    (referred to as 'origin') as well as the main ManimCommunity repository
+    (referred to as 'upstream').
```
```suggestion
    (referred to as 'community').
```

Aathish04 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+Note: if you are contributing new features or bug fixes, it is strongly
+preferred that you include tests that accompany them in your PR.  For this
+purpose, you can create a new file under the `tests/` folder and make sure that
+your new tests pass before submitting your PR.
```
```suggestion
your new tests pass before submitting your PR. Please check out the [Wiki Entry for Testing](https://github.com/ManimCommunity/manim/wiki/Testing) to learn more about how our tests work. 
```

Link them to the Wiki for further reading.

Aathish04 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+
+Note: if you are contributing documentation, you must also make sure that the
+command `make doctest`, executed from the `docs/` folder terminates without
+problems.
```
```suggestion
problems. Please check out the [Wiki Entry for Documentation](https://github.com/ManimCommunity/manim/wiki//Documentation-guidelines-(WIP)) to learn more about how we prefer our code documented.
```

Link them to the Wiki for further reading.

Aathish04 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+1. When submitting a PR, please make special note of whether your proposed
+   changes will result in breaking changes.
+
+1. When submitting a PR, make sure that your proposed changes are as general as
+   possible, and ready to be taken advantage by all of manim's users.  In
+   particular, leave out any machine-specific configurations, or any personal
+   information it may contain.
+
+2. If you are a maintainer, please label issues and PRs appropriately and
+   frequently.
+
+3. When opening a new issue, if there are old issues that are related, link
+   them in your new issue (even if the old ones are close).
+
+4. If you find stale or inactive issues that seem to be irrelevant, please post
+   a comment saying 'This issue should be closed', and a community developer
+   will take a look.
+
+5. Please do as much as possible to keep issues, PRs, and development in
+   general as tidy as possible.
```
```suggestion
2. When submitting a PR, make sure that your proposed changes are as general as
   possible, and ready to be taken advantage of by all of manim's users.  In
   particular, leave out any machine-specific configurations, or any personal
   information it may contain.

3. If you are a maintainer, please label issues and PRs appropriately and
   frequently.

4. When opening a new issue, if there are old issues that are related, link
   them in your new issue (even if the old ones are closed).

5. If you find stale or inactive issues that seem to be irrelevant, please post
   a comment saying 'This issue should be closed', and a community developer
   will take a look.

6. Please do as much as possible to keep issues, PRs, and development in
   general as tidy as possible.
```

Renumbered the points; fixed a typo.

huguesdevimeux commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+   your best Google practices.  If you can't find anything helpful, then go to
+   the next step.
+
+1. Make sure you can reproduce the issue, i.e. that you have some code that
```
Issue : Indexation mistake

leotrs replied:
You can actually use any numbers for lists in markdown, and they will be displayed correctly:

1. this list
1. has all indices
1. equal to one

PgBiel replied:
i mean, sure, but as a md file this looks weird and inconsistent, like, either 1. 1. 1. 1. or 1. 2. 3. 4. ,  not 1. 1. 2. 3. 4. as that has no apparent pattern

leotrs replied:
Ok, done!

huguesdevimeux replied:
> You can actually use any numbers for lists in markdown, and they will be displayed correctly:

TIL that :D

leotrs commented:
@huguesdevimeux already added a couple mentions to the wiki, as per @Aathish04 request.

eulertour commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+   differs from your expectation.
+
+3. Make sure to gather information about your environment, such as your
+   operating system, manim version, python version, and any stack traces that
```
I think it's best to just say to make sure you're using the latest version; I don't think we should support multiple versions or anything like that.

kolibril13 replied:
I am here with eulertour

leotrs replied:
Done

eulertour commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+3b1b](https://github.com/3b1b/manim), contributing in this way can be a bit
+confusing.  Here is a short guide on how to do it.
+
+1. First, make a fork of this repository.  This creates your own copy of the
```
Do you think it'd be better to just link to a tutorial on how to do this? I get the feeling there must be some tutorial out there which is more comprehensive, and it'd save us the trouble of answering issues on people's git setups.

leotrs replied:
How about this one? https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow. We can delete a good chunk of our document in favor of that one instead.

eulertour commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+   You can execute it by running `black <your_file>`.  This will make changes
+   to your code, and you will need to add/commit those changes.
+
+6. To request the ManimCommunity dev team to incorporate the changes in your
```
Should this be a 9?

leotrs replied:
Done

naveen521kk commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
-   git fetch fork
-   git checkout -b <your-branch-name> fork/master
+   git remote add upstream https://github.com/ManimCommunity/manim.git
+   git pull upstream master
```
You missed the checkout command. I think the user need to create a new branch locally before actually pushing it.
```suggestion
   git pull upstream master
   git checkout -b <your-branch-name> upstream/master
```

leotrs replied:
Ah good point. Done

kolibril13 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+5. Please give the community a reasonable amount of time before asking again,
+   or insisting on your issue.  Keep in mind that everyone is a volunteer.  If
+   you wait for a reasonable amount of time and you receive no response, feel
+   free to insist.
```
Suggestion: "feel free to insist." 
to "feel free to ask again"
sounds maybe a little more gentle.

leotrs replied:
Done

kolibril13 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+7. Further, make sure that you run the tests locally and that they all pass
+   before submitting a PR.  You can run the tests locally by executing `pytest`
+   from the main `manim/` folder (do not execute pytest from the code folder
+   `manim/manim`).
```
Is this really necessary? 
I normally don't do this.
I thought that github tests are there for this exact reason, that you don't have to run them locally.
Especially for new contributers, I think it would be nice to keep it as simple as possible

huguesdevimeux replied:
Hmmm I disagree here. Asking people to test locally will avoid 300 commits trying to fix tests on GitHub action.

leotrs replied:
> Is this really necessary? I normally don't do this.

You should!

> I thought that github tests are there for this exact reason, that you don't have to run them locally.

No. github tests run the tests on many platforms and many versions of different things. The author of a PR should be expected to run tests locally.

> Especially for new contributers, I think it would be nice to keep it as simple as possible

I assume you mean "inexperienced contributors" and not "new contributors". A new contributor to manim may already have a lot of experienced with open source and not be intimidate by local testing. Further, I don't think that asking people to run a single command with a single word (`pytest`) is asking too much.

kolibril13 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
-   Then you can make a Pull Request to the Manim Community Repo from your fork, through GitHub. Make sure to select `ManimCommunity/manim` instead of `3b1b/manim` as the `base repository` and your fork and branch as `head repository` - see the picture below.
+8. Finally, make sure to use the `black` autoformatter before submitting a PR.
+   You can execute it by running `black <your_file>`.  This will make changes
+   to your code, and you will need to add/commit those changes.
```
For me, this is a step that is a little annoying, primary because there is no good IDE support yet (I am using pycharm, maybe this is easier for other IDEs?) . Especially when you have more then one file, making this with every single file can get quit annoying. But maybe this is a point we can discuss in a separate issue. For now I am fine to leave it like this and get this pull request merged with this procedure.

kilacoda replied:
You can just run `black .` in the manim directory and it'll autoformat all files there, if they need to be changed; and given that the whole library has been preformatted, only the files you've changed will get formatted.

huguesdevimeux replied:
Agree with @kilacoda . Running `black .` (on `manim/`) is enough.

leotrs replied:
I agree that remembering to run black is annoying. But if you remember (for example, if it is explicitly described in the uigdelines! ;) ), running the command `black manim/` is fast and easy. There's literally no input necessary from the contributor/user.

kolibril13 commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,150 @@
+   will take a look.
+
+6. Please do as much as possible to keep issues, PRs, and development in
+   general as tidy as possible.
```
Here it would be nice to also add a list of these nice tags that @leotrs  suggested for code reviews like nitpicking, suggestion, etc.

leotrs replied:
Good catch. Added link

kolibril13 commented:
@leotrs : As you said "#311 is holding up another PR I have in store", I would be fine with it when we merge it like this for now.
However, I find the contribution guidelines introduce a lot of barriers for new contributors, this can be discussed in #319 , and then this document could be adjusted in a future pull request.

Further, we could also introduce some guidelines, how pull requests are handled by others.
e.g.:
* at least 2 manim-devs are reviewing the code
* After running github tests online, other people will test the code (here a question: Should code be run and tested locally at the reviewer's computers, or is it ok to just have a look online, if the code looks reasonable?)
* For respectful communication, we can use these tags: nitpicking, suggestion, etc...

kilacoda commented:
> Further, we could also introduce some guidelines, how pull requests are handled by others.
> e.g.:
> 
> * at least 2 manim-devs are reviewing the code
> * After running github tests online, other people will test the code (here a question: Should code be run and tested locally at the reviewer's computers, or is it ok to just have a look online, if the code looks reasonable?)
> * For respectful communication, we can use these tags: nitpicking, suggestion, etc...

Aren't we.... already doing those things?

kolibril13 commented:
> Aren't we.... already doing those things?

Yes. My intention is: Someone who wants to contribute, will then know how we handle his/her pull request.

huguesdevimeux commented:
> Further, we could also introduce some guidelines, how pull requests are handled by others.

I agree with that. It would be nice to explain how we work so contributor can adapt and make the work easier !

leotrs commented:
> * at least 2 manim-devs are reviewing the code
> * After running github tests online, other people will test the code (here a question: Should code be run and tested locally at the reviewer's computers, or is it ok to just have a look online, if the code looks reasonable?)
> * For respectful communication, we can use these tags: nitpicking, suggestion, etc...


Good point, I added some language to this effect.

kilacoda commented:
Suggestion: Can we make it a requirement to fill out the PR form properly? It takes a bit of time, but makes things clearer imo.

leotrs commented:
@kilacoda Good point. Just added that to the latest commit.

Also, after the discussion in #319, I added a note saying that not all of those pre-requisites are strictly mandatory, but that it is strongly encouraged to do them. It also mentions WIPs and drafts.

Imma leave this open a couple more days to see if there are more comments and then merge. Thanks for your eyes everyone. This is a very important document as it is part of the public face of manim.

naveen521kk commented on CONTRIBUTING.md:
````diff
@@ -1,51 +1,193 @@
- ```
+   ```sh
+   git remote add upstream https://github.com/ManimCommunity/manim.git
+   git pull upstream master
````
**Issue**: It should be `git fetch` and not `git pull`. `git pull` will create a merge commit which is unnecessary hear and could cause problems.
```suggestion
   git fetch upstream
```

naveen521kk commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,193 @@
+6. If you want your changes to be incorporated to the main ManimCommunity
+   repository, you need to make sure that there are no merge conflicts between
+   the current upstream/master and the changes you are trying to make.  For
+   this purpose, please execute `git pull origin master` before you push
```
Isn't it `git pull upstream master`?
```suggestion
   this purpose, please execute `git pull upstream master` before you push
```
This statement mentioned in original code will just pull the master branch of their fork which is unnecessary, it should pull from master branch on manim community.

naveen521kk commented on CONTRIBUTING.md:
```diff
@@ -1,51 +1,193 @@
-1. First, make a fork of this repository.
-2. Then clone the repository you forked using the command below (bash/zsh).
+3. You have to make `git` be aware of the three repositories, by entering the
+   commands below.
```
Isn't it two repos?

leotrs replied:
Fair

naveen521kk reviewed (approved):
Looks good to me. :)
